main.c

```diff
@@ -0,0 +1,440 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This program presents a menu of installation ISOs that we can possibly
+ * chain-boot to.  It uses JSON information obtained from SimpleStreams to
+ * provide the list of ISOs with a friendly label.
+ *
+ * The menu is styled to have an appearance that is as close to Subiquity as
+ * possible.
+ *
+ * Input is a file similar to
+ * http://cdimage.ubuntu.com/streams/v1/com.ubuntu.cdimage.daily:ubuntu-server.json
+ *
+ * The chosen ISO is output in a format friendly for the /bin/sh source
+ * built-in - sample output:
+ *
+ * MEDIA_URL="https://releases.ubuntu.com/kinetic/ubuntu-22.10-live-server-amd64.iso"
+ * MEDIA_LABEL="Ubuntu Server 22.10 (Kinetic Kudu)"
+ * MEDIA_SIZE="1642631168"
+ */
+
+#define _GNU_SOURCE
+#include <locale.h>
+#include <ncurses.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/param.h>
+
+#include <json-c/json.h>
+
+int ubuntu_orange = COLOR_RED;
+int text_white = COLOR_WHITE;
+int back_green = COLOR_GREEN;
+
+void usage(char *prog)
+{
+    printf("usage: %s --input path --output path", prog);
+    exit(1);
+}
+
+typedef struct _args_t
+{
+    char *infile;
+    char *outfile;
+} args_t;
+
+args_t *args_create(int argc, char **argv)
+{
+    int cur = 1;
+    args_t *args = calloc(sizeof(args_t), 1);
+
+    while(cur < argc) {
+        if(!strcmp(argv[cur], "--input") && cur + 1 < argc) {
+            args->infile = strdup(argv[cur + 1]);
+            cur += 2;
+        } else if(!strcmp(argv[cur], "--output") && cur + 1 < argc) {
+            args->outfile = strdup(argv[cur + 1]);
+            cur += 2;
+        } else {
+            usage(argv[0]);
+        }
+    }
+    if(!args->infile || !args->outfile) {
+        usage(argv[0]);
+    }
+
+    return args;
+}
+
+void args_free(args_t *args)
+{
+    if(!args) return;
+
+    free(args->infile);
+    free(args->outfile);
+    free(args);
+}
+
+void logger(char *fmt, ...)
+{
+    va_list ap;
+    va_start(ap, fmt);
+
+    FILE *out = fopen("/tmp/iso-chooser-menu.log", "a");
+    vfprintf(out, fmt, ap);
+    fprintf(out, "\n");
+    fclose(out);
+}
+
+char *saprintf(char *fmt, ...)
+{
+    va_list ap;
+    va_start(ap, fmt);
+
+    char *ret = NULL;
+    vasprintf(&ret, fmt, ap);
+    return ret;
+}
+
+typedef enum {
+    DECREASE=-1,
+    SELECT=0,
+    INCREASE=1,
+} choice_event;
+
+typedef struct _iso_data_t
+{
+    char *label;
+    char *url;
+    int size;
+} iso_data_t;
+
+iso_data_t *iso_data_create(char *label, char *url, int size)
+{
+    iso_data_t *ret = calloc(sizeof(iso_data_t), 1);
+    if(!ret) return NULL;
+
+    ret->label = label;
```

Comment:
It feels weird to me to steal `label` and `url` from the caller while also expecting the memory to be allocated using malloc/calloc. Would you consider `strdup`ing here instead of in the caller?

Reply:
The one hardcoded case now is to be removed, so long term I expect that the caller has already allocated memory.  I think I prefer to document my memory assumptions here.

Reply:
Could also do something like this:
```
    choices->values[1] = iso_data_create(
            "Server",
            json_object_get_string(title),
            json_object_get_string(codename)),
            json_object_get_string(path)),
            json_object_get_int(size));
```
and let iso_data_create() generate the strings, but in general I wanted the choices_t to own the memory so that choices_free() would handle it without issue.

Reply:
It feels super dirty but I think it would be a defensible choice for this program to never deallocate anything!

Reply:
(exit_group(2) hides all memory management sins!)

Reply:
I'm happy to mark this discussion resolved if needed.